config-controller/pkg/client/client.go

```diff
@@ -246,36 +247,56 @@ func (c *client) CreatePolicy(ctx context.Context, policy *storage.Policy) (*sto
 		return nil, errors.Wrap(err, fmt.Sprintf("Failed to POST policy '%s'", policy.Name))
 	}
 
-	c.cache[createdPolicy.Name] = createdPolicy
+	c.policyObjectCache[createdPolicy.GetId()] = createdPolicy
+	c.policyNameToIDCache[createdPolicy.GetName()] = createdPolicy.GetId()
 
 	return createdPolicy, nil
 }
 
 func (c *client) UpdatePolicy(ctx context.Context, policy *storage.Policy) error {
 	getLogger(ctx).Info("PUT", "policyName", policy.Name)
-	err := c.svc.PutPolicy(ctx, policy)
 
+	existingPolicy := c.policyObjectCache[policy.GetId()]
+	// Do not update/overwrite if name collision with imperative policies
+	if id, ok := c.policyNameToIDCache[policy.GetName()]; ok {
+		if existingPolicy.GetSource() != storage.PolicySource_DECLARATIVE {
```

Comment:
`existingPolicy` would be `nil` if this is the first time reconciling this CR.

Reply:
But we would not enter the if case then, since we would not find the entry for this policy in the policy name to id cache. But I agree, will rework this a bit for better readability and flow.